## SCRDR Tagging Experiment

ဒီ notebook ကတော့ SCRDR ကိုသုံးပြီး POS Tagging experiment ကို လက်တွေ့ run ပြထားတာ ဖြစ်ပါတယ်။  

ဒီမှာ သုံးပြထားတဲ့ scrdr_tagger.py code က အချိန်တော်တော်ပေးပြီး ရေးဖြစ်ခဲ့ပါတယ်။ အလုပ်အားတဲ့အချိန်မှာ ရေးလိုက် experiment လုပ်ကြည့်လိုက်နဲ့ စုစုပေါင်း သုံးလခန့်ကြာတယ်လို့ မှတ်မိနေပါတယ်။ POS tagging နဲ့ word tokenizer ကလည်း အခြေခံအားဖြင့် tagging လုပ်တာပဲ၊ အတူတူပဲလို့ နားလည်တာမို့ အခု ရှဲပေးထားသလိုမျိုး ပရိုဂရမ်နှစ်ပုဒ် မဟုတ်ပဲ ပရိုဂရမ် တစ်ပုဒ်ထဲ အနေနဲ့ ပေါင်းပြီး ရေးတာမျိုးလည်း လုပ်ကြည့်ဖြစ်ခဲ့ပါတယ်။ သို့သော်လည်း လက်တွေ့ function တွေကို ရေးတဲ့အခါမှာ တထပ်တည်းမတူတာမို့လို့ ကိုယ်ရေးထားတဲ့ ကုဒ်ကို ပြန်ကြည့်ပြီး ရှုပ်တယ်လို့၊ တခြားလိုက်ဖတ်တဲ့ သူတွေအတွက် ပိုခက်နိုင်တယ်လို့ ယူဆပြီး နောက်ဆုံးမှာ နှစ်ပုဒ် သပ်သပ်စီပဲ ခွဲရေးလိုက်ပါတယ်။  

Vibe coding ကိုလည်း Deepseek, ChatGPT, Gemini, Claude တွေနဲ့ လိုတဲ့အပိုင်းကို စမ်းရေးခိုင်းလိုက်၊ ရေးပြီးထွက်လာတဲ့ ကုဒ်ကို ဖတ်ကြည့်လိုက်၊ run ကြည့်လိုက်လုပ်ပြီးလည်း အမျိုးမျိုး လေ့လာဖြစ်ခဲ့ပါတယ်။ ဘာသွားတွေ့ရသလဲ ဆိုတော့ Ripple Down Rules နဲ့ ပတ်သက်တဲ့ ကုဒ်ကို အစအဆုံး အဆင်ပြေအောင် ရေးပေးနိုင်တာမျိုးကို လက်ရှိ LLM တွေအနေနဲ့က တော်တော်လေး ခက်ခဲတာကို သွားတွေ့ရပါတယ်။ LLM တွေထဲမှာတော့ SCRDR အတွက် Claude ကပဲ အကောင်းဆုံးလို့ ပြောရပါလိမ့်မယ်။ Module တွေအနေနဲ့ခွဲရေးလိုက် မြန်မာစာ myPOS ဒေတာကို သုံးပြီးစမ်းလိုက်၊ ရလဒ်က မကောင်းလို့ ပြန်ပြင်လိုက်၊ ရလဒ်က မဆိုးပေမဲ့လည်း run time က အရမ်းကြာနေလို့ debugging လုပ်လိုက်နဲ့ လက်ရှိဗားရှင်း ရောက်အောင်၊ ကိုယ်လိုချင်တဲ့ အတိုင်းအတာ တစ်ခုထိ ရောက်အောင် တော်တော် အချိန်ပေးလိုက်ရပါတယ်။ နောက်ဆုံးမှာတော့ Claude နဲ့ပဲ refine ပြန်လုပ်ထားတာပါ။  

Refer လုပ်ခဲ့တဲ့ code က အောက်ပါ Github repository ပါ။  
(ဒီ code ကိုသုံးပြီး ၂၀၁၇ က ကျင်းပခဲ့တဲ့ CICLING 2017 conf. အပါအဝင်၊ စာတမ်း သုံးလေးစောင်လောက် ဖတ်ခဲ့ပါသေးတယ်။)  

[https://github.com/datquocnguyen/RDRPOSTagger](https://github.com/datquocnguyen/RDRPOSTagger)  

ဒီ scrdr_tagger.py ကုဒ်က NLP tagging task တွေအတွက်၊ သုတေသန အလုပ်တွေအတွက် အသုံးဝင်မယ်လို့ ယုံကြည်နေပါတယ်။  

Ye Kyaw Thu 
Lab Leader,  
Language Understanding Lab, Myanmar.  
13 April 2026  

In [1]:
%pwd

'/home/ye/exp/scrdr_tuto/tagger'

In [3]:
!ls ./pos_data/*

./pos_data/corpus.txt	  ./pos_data/otest.txt
./pos_data/mk-wordtag.pl  ./pos_data/otest.word


In [4]:
%cd ./pos_data/

/home/ye/exp/scrdr_tuto/tagger/pos_data


In [5]:
!head ./corpus.txt

၁၉၆၂/num ခုနှစ်/n ခန့်မှန်း/v သန်းခေါင်စာရင်း/n အရ/ppm လူဦးရေ/n ၁၁၅၉၃၁/num ယောက်/part ရှိ/v သည်/ppm ။/punc
လူ/n တိုင်း/part တွင်/ppm သင့်မြတ်/v လျော်ကန်/v စွာ/part ကန့်သတ်/v ထား/part သည့်/part အလုပ်/n လုပ်/v ချိန်/n အပြင်/conj ၊/punc လစာ/n နှင့်တကွ/conj အခါ/n ကာလ/n အားလျော်စွာ/ppm သတ်မှတ်/v ထား/part သည့်/part အလုပ်/n အားလပ်ရက်/n များ/part ပါဝင်/v သည့်/part အနားယူခွင့်/n နှင့်/conj အားလပ်ခွင့်/n ခံစားပိုင်ခွင့်/n ရှိ/v သည်/ppm ။/punc
ဤ/adj နည်း/n ကို/ppm စစ်ယူ/v သော/part နည်း/n ဟု/part ခေါ်/v သည်/ppm ။/punc
စာပြန်ပွဲ/n ဆို/v တာ/part က/ppm အာဂုံဆောင်/v အလွတ်ကျက်/v ထား/part တဲ့/part ပိဋကတ်သုံးပုံ/n စာပေ/n တွေ/part ကို/ppm စာစစ်/v သံဃာတော်ကြီး/n တွေ/part ရဲ့/ppm ရှေ့/n မှာ/ppm အလွတ်/adv ပြန်/v ပြီး/part ရွတ်ပြ/v ရ/part တာ/part ပေါ့/part ။/punc
ဒီ/pron မှာ/ppm ကျွန်တော့်/pron သက်သေခံကတ်/n ပါ/part ။/punc
၂ဝ/num ရာစု/n မြန်မာ့/n သမိုင်း/n သန်းဝင်းလှိုင်/n ၊/punc ၂ဝဝ၉/num ခု/part ၊/punc မေ/n လ/n ၊/punc ကံကော်ဝတ်ရည်/n စာပေ/n ။/punc
ကျွန်တော်/pron မျက်မှန်/n တစ်/tn လက်/part လုပ်/v ချင်/part ပါ/p

## mk-wordtag Perl Code

word နဲ့ tag ကို ခွဲထုတ်ဖို့အတွက် ရေးထားတဲ perl script ပါ။  

In [6]:
!cat ./mk-wordtag.pl

#!/usr/bin/perl
use warnings;
use utf8;

#last updated: 16 May 2017
#written by Ye, AI Lab.,
#Okayama Prefectural University, Japan
#How to run: perl mk-wordtag.pl <input-file-name> <delimeter> <w|t|wt|cw|c> 
#Here, 
# w = print word only (i.e. without POS tags), 
# t = print tag only
# wt = print word/tag
# cw = print words including compound words,
# lcw = list compound words, 
# c = print sentence that contain tagging error of "word/" 
# 
# How to run:
# e.g. ./mk-wordtag.pl ./kh-pos.all.f2.utf8 "\/" w | less -r
# e.g ./mk-wordtag.pl ./kh-pos.all.f2.utf8 "\/" t
# e.g ./mk-wordtag.pl ./kh-pos.all.f2.utf8 "\/" wt

binmode STDIN,  ":utf8";
binmode STDOUT, ":utf8";

my $TagMarker=$ARGV[1]; # give command line parameter such as "\|", "\/" ...
my $word_or_tag=$ARGV[2];

open (my $inputFILE,"<:encoding(utf8)", $ARGV[0]) or die "Couldn't open input file $ARGV[0]!, $!\n";

my $one_token; my $tmpLine=""; my $tmpLine2="";

   while($line = <$inputFILE>)
   {
      if ($line!~/^$/)
      {
    

## Two Types of Test Dataset

Testing အတွက် ဒေတာကို format နှစ်မျိုး သတ်မှတ်ထားပါတယ်။ အောက်ပါ otest.txt က testing လုပ်ရင်းနဲ့ တခါတည်း evaluation ပါ လုပ်ပေးနိုင်ဖို့အတွက် သုံးမယ့် ဖိုင်ပါ။ ဆိုလိုတာက evaluation လုပ်ဖို့အတွက် လူကိုယ်တိုင် tag လုပ်ထားတဲ့ POS tag တွေပါ တွဲပါနေတဲ့ test data ပါ။  

In [7]:
!head ./otest.txt

တစ်/tn ကိုက်/n ကို/ppm ဝမ်/n ခုနှစ်ထောင်/tn ပါ/part ။/punc
မနှစ်/n က/ppm သူ/pron ကျွန်မ/pron ကို/ppm သင်/v ပေး/part တယ်/ppm ။/punc
ကျွန်တော့်/pron ခုံ/n သွား/v ရှာ/v မလို့/part ။/punc
အတန်း/n စ/v တာ/part ကြာ/v ပြီ/ppm လား/part ။/punc
ဆေး/n နည်းနည်း/adv စား/v လိုက်/part ၊/punc သုံး/tn လေး/tn ရက်/n လောက်/part အနားယူ/v လိုက်/part ရင်/conj ပျောက်/v သွား/part မှာ/ppm ပါ/part ။/punc
အေးချမ်း/v မှု/part နဲ့/conj စည်းကမ်း/n ကို/ppm တည်မြဲ/v အောင်/part ထိန်းသိမ်း/v သည်/ppm ။/punc
ဇွန်း/n ကို/ppm လိုအပ်/v တယ်/ppm ။/punc
ဘွဲ့/n ရ/v ရင်/conj ဘာ/n လုပ်/v မ/part လို့/part လဲ/part ။/punc
ကျွန်တော်/pron ချောင်းဆိုး/v ခြင်း/part အတွက်/ppm တစ်/tn ခု/part ခု/part လို/v ချင်/part တယ်/ppm ။/punc
အသီးအနှံ/n တို့/part မှ/ppm လွဲ/v လျှင်/conj လူ/n တို့/part ၏/ppm အဓိက/n အစားအစာ/n မှာ/ppm ငါး/n ဖြစ်/v သည်/ppm ။/punc


နောက်ထပ် တစ်မျိုးကတော့ အောက်မှာ ပြထားတဲ့ otest.word လိုမျိုး ပုံမှန် text ဖိုင် format ပါ။ ဒါကတော့ လက်တွေ့ raw data တွေနဲ့ POS tagging လုပ်နိုင်ဖို့အတွက် support လုပ်ပေးထားတဲ့ ပါ။    

In [8]:
!head ./otest.word

တစ် ကိုက် ကို ဝမ် ခုနှစ်ထောင် ပါ ။
မနှစ် က သူ ကျွန်မ ကို သင် ပေး တယ် ။
ကျွန်တော့် ခုံ သွား ရှာ မလို့ ။
အတန်း စ တာ ကြာ ပြီ လား ။
ဆေး နည်းနည်း စား လိုက် ၊ သုံး လေး ရက် လောက် အနားယူ လိုက် ရင် ပျောက် သွား မှာ ပါ ။
အေးချမ်း မှု နဲ့ စည်းကမ်း ကို တည်မြဲ အောင် ထိန်းသိမ်း သည် ။
ဇွန်း ကို လိုအပ် တယ် ။
ဘွဲ့ ရ ရင် ဘာ လုပ် မ လို့ လဲ ။
ကျွန်တော် ချောင်းဆိုး ခြင်း အတွက် တစ် ခု ခု လို ချင် တယ် ။
အသီးအနှံ တို့ မှ လွဲ လျှင် လူ တို့ ၏ အဓိက အစားအစာ မှာ ငါး ဖြစ် သည် ။


## Python Code for RDR based Tagging  

scrdr_tagger.py code ကို လေ့လာကြည့်ကြရအောင်။  

In [9]:
%cd ..

/home/ye/exp/scrdr_tuto/tagger


In [14]:
!cat ./scrdr_tagger.py

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
  RDR Tagger  —  Single Classification Ripple Down Rules based NLP Sequence Tagger

A complete, production-grade reimplementation of the SCRDR-based POS / NLP
sequence tagger described in:

  Nguyen et al. (2014). "A Robust Transformation-Based Learning Approach Using
      Ripple Down Rules for Part-of-Speech Tagging."  arXiv:1412.4021
  Nguyen & Widyantoro (2014). "RDRPOSTagger: A Ripple Down Rules-based
      Part-Of-Speech Tagger."  EACL 2014.

Vibe coding by Ye Kyaw Thu, Lab. Leader, Myanmar.
Last updated: 13 April 2026.

Key improvements over the original RDRPOSTagger codebase
---------------------------------------------------------
  1. NO eval() / exec()  — All condition checking uses structured tuples with
       direct equality comparison.  This is 4–10× faster and safe.
  2. Multiprocessing learning  — Each initial tag group builds its subtree in a
       separate worker process using multiprocessing.Pool, with shared read-

## Shell Script

Experiment လုပ်တဲ့အခါမှာ သုံးတဲ့ shell script က အောက်ပါအတိုင်းပါ။  

In [11]:
!cat run_all.sh

#!/bin/bash

## for training:
time python ./scrdr_tagger.py train --train ./pos_data/corpus.txt --model model/mypos \
--jobs 16 --threshold-imp 2 --threshold-match 2

## for testing:  
time python ./scrdr_tagger.py test --test ./pos_data/otest.txt --model ./model/mypos \
--hyp ./otest.hyp --confusion-matrix pos_cm.png

## for tagging:
time python ./scrdr_tagger.py tag --input ./pos_data/otest.txt --model ./model/mypos \
--output ./otest.tag

## Evaluation
python2.7 ./evaluate.py ./otest.hyp ./pos_data/otest.txt | tee eval.log


## --help

Commandline argument တွေကို လေ့လာကြည့်ကြရအောင်။  

In [19]:
!python ./scrdr_tagger.py --help

usage: rdr_tagger [-h] [--log-level {DEBUG,INFO,WARNING,ERROR}]
                  {train,test,tag} ...

  RDR Tagger  —  Single Classification Ripple Down Rules based NLP Sequence Tagger

A complete, production-grade reimplementation of the SCRDR-based POS / NLP
sequence tagger described in:

  Nguyen et al. (2014). "A Robust Transformation-Based Learning Approach Using
      Ripple Down Rules for Part-of-Speech Tagging."  arXiv:1412.4021
  Nguyen & Widyantoro (2014). "RDRPOSTagger: A Ripple Down Rules-based
      Part-Of-Speech Tagger."  EACL 2014.

Vibe coding by Ye Kyaw Thu, Lab. Leader, Myanmar.
Last updated: 13 April 2026.

Key improvements over the original RDRPOSTagger codebase
---------------------------------------------------------
  1. NO eval() / exec()  — All condition checking uses structured tuples with
       direct equality comparison.  This is 4–10× faster and safe.
  2. Multiprocessing learning  — Each initial tag group builds its subtree in a
       separate worker 

## POS Tagging

myPOS ဒေတာကို သုံးပြီး POS tagging experiment လုပ်ကြည့်ကြရအောင်။  

In [12]:
!./run_all.sh

15:16:15 [INFO] ============================================================
15:16:15 [INFO] MODE: TRAIN
15:16:15 [INFO]   Train corpus : ./pos_data/corpus.txt
15:16:15 [INFO]   Model prefix : model/mypos
15:16:15 [INFO]   Imp thresh   : 2
15:16:15 [INFO]   Match thresh : 2
15:16:15 [INFO]   Workers      : 16
15:16:15 [INFO] ============================================================
15:16:15 [INFO] Corpus: 43196 sentences, 564517 tokens
15:16:16 [INFO] Lexicon built: 34775 entries | unk_word=part | unk_cap=fw | unk_num=fw
15:16:16 [INFO] Lexicon saved → model/mypos.lex
15:16:16 [INFO] Building object dictionary from 43196 sentences …
15:16:17 [INFO] Object dictionary: 15 initial tags, 564517 total tokens  [1.8s]
15:16:17 [INFO] Default tag (root conclusion): part
15:16:17 [INFO] Learning exception rules for 15 initial tags using 15 worker(s) …
15:18:35 [INFO] Rule learning complete  [137.5s]
15:18:35 [INFO]   Tag abb          : 1 exception rule(s)
15:18:35 [INFO]   Tag adj          :

## Let's Peek SCRDR POS Tagging Model

Training လုပ်ပြီးရလာတဲ့ SCRDR Tree ကို ဝင်လေ့လာကြည့်ကြရအောင်။  

In [13]:
!head -n 300 ./model/mypos.rdr

True : part
	tag == "abb" : abb
		prevTag1 == "punc" and suffixL2 == "<SFX>" : fw
	tag == "adj" : adj
		nextWord1 == "မှာ" : pron
			nextTag1 == "ppm" : adj
				wordShape == "U" : v
					nextWord1 == "မှာ" : n
		nextTag1 == "ppm" and prefixL2 == "ဒီ" : pron
			nextTag1 == "ppm" : adj
				suffixL3 == "<SFX>" : n
		nextTag1 == "ppm" and prefixL2 == "ဘယ" : pron
		nextWord1 == "သို့" : pron
			nextTag1 == "ppm" : adj
				prevWord1 == "အခြား" : n
		nextWord1 == "လို" : pron
			nextTag1 == "v" and wordShape == "U" : adj
				nextWord1 == "လို" and nextWord2 == "ချင်" : n
		word == "ဘယ်" and nextTag1 == "part" : pron
			nextTag1 == "part" and prefixL2 == "ဘယ" : adj
				prevWord1 == "ပဲ" : adv
		word == "ဒီ" and nextTag1 == "part" : pron
			suffixL3 == "<SFX>" : adj
		prevTag2 == "<BOS_T>" and prevTag1 == "<BOS_T>" and word == "အားလုံး" : pron
			prevTag2 == "<BOS_T>" and prevTag1 == "<BOS_T>" : n
				prevTag2 == "<BOS_T>" and prevTag1 == "<BOS_T>" and word == "အားလုံး" : adj
		nextTag1 == "v" 

## Evaluation with evaluate.py Code

2017 တုန်းက တင်ထားခဲ့တဲ့ CICLING 2017 စာတမ်းမှာတုန်းက သုံးခဲ့တဲ့ evaluate.py ကိုပဲ သုံးပြီး evaluation လုပ်ကြည့်ပါမယ်။ အဲဒါမှ fair comparison ဖြစ်မှာမို့။ တစ်ခုရှိတာက လက်ရှိ training လုပ်ဖို့အတွက် သုံးထားတဲ့ myPOS corpus က version 3 မို့လို့ 2017 တုန်းက သုံးခဲ့တဲ့ myPOS corpus ထက် သုံးဆ ကျော်ကွာပါတယ်။  

In [15]:
%pwd

'/home/ye/exp/scrdr_tuto/tagger'

In [17]:
!python2.7 ./evaluate.py otest.hyp ./pos_data/otest.txt

Tag precision: 0.945426195426
Recall: 0.945426195426
F-Measure:  0.998217998218


CICLING စာတမ်းမှာတုန်းက CRF, HMM စတဲ့ ML method ၅မျိုးနဲ့ RDR ကို နှိုင်းယှဉ်ခဲ့ပါတယ်။ အဲဒီတုန်းက လုပ်ခဲ့တဲ့ experiment တွေရဲ့ ရလဒ်က အောက်ပါအတိုင်းပါ။ Open test ဒေတာနဲ့ ရတဲ့ RDR ရဲ့ ရလဒ်က 97.05% ပါ။    

|Methods|Closed|Test-set|Open Test-set|  
| ---| --- | --- | --- |
|CRFs|97.77%|95.05%|  
|HMM|97.31%|96.43%|  
|MaxEnt|96.55%|96.31%|  
|RDR|98.42%|**97.05%**|  
|SVM|**99.83%**|93.55%|  
|Two-Hours|95.83%|92.87%|  

အဲဒါကြောင့် လက်ရှိ scrdr_tagger.py နဲ့ myPOS (version 3.0) ဒေတာနဲ့ ရလဒ်ဖြစ်တဲ့ F-Measure:  0.998217998218 က ပြတ်ပြတ်သားသား သာပါတယ်။   

## References

ကျွန်တော်တို့ Language Understanding Lab. အနေနဲ့ SCRDR ကို သုံးပြီး experiment တွေအများကြီး လုပ်ဖြစ်ခဲ့ကြပါတယ်။ SCRDR ကိုသုံးခဲ့တဲ့ စာတမ်း တချို့ကို လက်လှမ်းမီသလောက် reference အနေနဲ့ ထည့်ပေးထားလိုက်ပါတယ်။  
- RDRPOSTagger: [https://github.com/datquocnguyen/RDRPOSTagger](https://github.com/datquocnguyen/RDRPOSTagger)  
- Khin War War Htike, Ye Kyaw Thu, Zuping Zhang, Win Pa Pa, Yoshinori Sagisaka and Naoto Iwahashi, "Comparison of Six POS Tagging Methods on 10K Sentences Myanmar Language (Burmese) POS Tagged Corpus", at 18th International Conference on Computational Linguistics and Intelligent Text Processing (CICLing 2017), April 17~23, 2017, Budapest, Hungary.
- Khaing Hsu Wai, Ye Kyaw Thu, Swe Zin Moe, Hnin Aye Thant, Thepchai Supnithi, "Myanmar (Burmese) String Similarity Measures based on Phoneme Similarity", Journal of Intelligent Informatics and Smart Technology, April 1st Issue, 2020, pp. 27-34. (submitted December 21, 2019; accepted March 6, 2020; revised March 16, 2020; published online April 30, 2020) 
- Wint Theingi Zaw, Shwe Sin Moe, Ye Kyaw Thu, Nyein Nyein Oo, "Applying Weighted Finite State Transducers and Ripple Down Rules for Myanmar Name Romanization", In Proceedings of the 17th International Conference on Electrical Engineering/Electronics, Computer, Telecommunications and Information Technology (ECTI-CON 2020), June 24-27, 2020, Virtual Conference Hosted by College of Computing, Prince of Songkla University, Thailand, pp. 143-148.
- Wint Theingi Zaw, Ye Kyaw Thu, Shwe Sin Moe, Nyein Nyein Oo, Thepchai Supnithi, "Statistical Machine Translation, Ripple Down Rules and Hidden Markov Model for Burmese Romanization", In Proceedings of The 4th International Conference on Advanced Information Technologies (ICAIT-2020), Nov 4-5 2020, Yangon, Myanmar, pp. 99-104.
